Japanese OTO Template Generators

adlez27 opened the pull request:
# Tempo-based VCV
This is a new generator, separate from the existing VCV template generator.  
Depending on community feedback it may replace the original.

- Removed user inputs for overlap, preutterance, fixed, and cutoff. These are calculated from the tempo instead.

# CVVC
Substantial changes have been made to the original CVVC template generator for better base parameters. The previous behavior resulted in OTOs that required many major batch edits before it could be fine-tuned for the voicebank.

- Removed user inputs for CV overlap/preutterance/fixed/cutoff and VC preutterance/fixed/cutoff
- Added user input for initial vowel stabilization (distance from preutterance to fixed)
- OTO parameters are now calculated from tempo, consonant length, and VC overlap
- Unique logic for all of these OTO types: head CV, head V, CV, VC, solo C, solo V, VV, VCV, tail, other
- Entry tags use capitalization to distinguish similar names for ease of filtering
- Fixed bug with append suffix, now generating the expected tail OTOs

Example screenshots (using synthesized waveforms for clarity)

## Entry list
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/bd6ad811-89ed-42be-a040-3808a12a64eb)
After (Missing tail entries included, solo V entries included)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/2e9cada9-180e-4777-ab49-4ff9576770b2)

## Head CV
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/65717b5f-80ee-44ce-8d47-54b1feb4cf58)
After (Overlap is not needed when there are no preceding notes)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/f118cbd5-98bc-447c-ad3f-81cdece75862)

## Head V
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/490eed87-0d9c-47b2-8412-c2d99b2bf7b5)
After (Removed unnecessary silence, and overlap is not needed when there are no preceding notes)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/80360c67-1e23-4b72-8fd3-8d83854266d2)

## CV
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/6626cdc4-a333-488b-884f-51b64dfee8f4)
After (More cutoff to include the rest of the vowel, since the generator knows there's no consonant after this entry)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/7dba876a-3e81-4732-82ea-580f544ead75)

## VC
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/4af3aa7b-0e05-4da5-9e13-bf976f3c07ff)
After (Parameters are calculated from VC overlap and consonant length, not manually specified by user)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/316405a2-a85e-490c-a472-b636c52c7859)

## Solo C
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/9abfd924-b3be-49b1-96f2-3f8bb94dd08b)
After (Unique logic for consonant-only OTO, instead of reusing VC OTO)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/f735a5d3-0be9-4630-b0c0-a555f3658f2a)

## Solo V
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/b4060a0d-c322-440a-83a7-398cc258e807)
After (Unique logic for vowel-only OTO, instead of reusing CV OTO)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/5dd45cad-2072-4039-be50-0ae8568de6a5)

## VV
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/a74f1e08-9ae1-45c8-9e59-33abe462851f)
After (Now using VC overlap for previous vowel, instead of reusing CV OTO)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/8dcb2370-b93b-4c93-a6d9-8d70b175bbaf)

## VCV
Before
![image](https://github.com/sdercolin/vlabeler/assets/26780012/4ea37031-5872-421c-b8c4-a6bed80ad296)
After (Proper VCV OTO, instead of reusing CV OTO)
![image](https://github.com/sdercolin/vlabeler/assets/26780012/0d553f55-2820-441b-bb91-bf73f54895a6)

## Tail
![image](https://github.com/sdercolin/vlabeler/assets/26780012/094348a4-2407-458a-8aa5-a114c71df540)

sdercolin commented:
@adlez27 Thank you for the PR!

I have some questions before the detailed review: 
1. How was the relationship(ratio) of different parameters decided?
2. Is this algorithm adapted for all types of voicebanks? (For example, may some users have special recording habits so that the algorithm is not fitting?)
3. For VCV it's added as a new plugin for testing, what's the reason that for CVVC, the original one is edited?

sdercolin commented:
Oh, and could you target the `dev` branch instead?

adlez27 commented:
1. In all of these descriptions, start = initial offset + index * beat length, which is the preutterance location for that index's CV. The parameter description is for the location on the wave file.
    - Head CV
        Offset: start - consonant length - 10
        Overlap: start - consonant length
        Preutterance: start
        Fixed: start + vowel stabilizing length
        Cutoff: next start - VC overlap (- consonant length, if next has consonant)
    - Head V
        Offset: start - 20
        Overlap: start - 10
        Preutterance: start
        Fixed: start + vowel stabilizing length
        Cutoff: next start - VC overlap (- consonant length, if next has consonant)
    - CV
        Offset: start - consonant length
        Overlap: start - consonant length/2
        Preutterance: start
        Fixed: start + vowel stabilizing length
        Cutoff: next start - VC overlap (- consonant length, if next has consonant)
    - VC
        Offset: start - consonant length - 2 * VC overlap
        Overlap: offset + VC overlap
        Preutterance: offset + 2 * VC overlap
        Fixed: preutterance + 10
        Cutoff: start
    - Solo C
        Offset: start - consonant length
        Overlap: consonant length/2
        Preutterance: consonant length/2
        Fixed: 0
        Cutoff: start
    - Solo V
        Offset: start + vowel stabilizing length
        Overlap: offset + VC overlap
        Preutterance: offset + VC overlap
        Fixed: 0
        Cutoff: next start - VC overlap (- consonant length, if next has consonant)
    - VV
        Offset: start - 2 * VC overlap
        Overlap: offset + VC overlap
        Preutterance: start
        Fixed: preutterance + vowel stabilizing length
        Cutoff: next start - VC overlap (- consonant length, if next has consonant)
    - VCV
        Offset: start - consonant length - 2 * VC overlap
        Overlap: offset + VC overlap
        Preutterance: start
        Fixed: preutterance + vowel stabilizing length
        Cutoff: next start - VC overlap (- consonant length, if next has consonant)
    - Tail
        Offset: start - 2 * VC overlap
        Overlap: offset + VC overlap
        Preutterance: start
        Fixed: preutterance + 10
        Cutoff: preutterance + 20
    - Other
        Offset: a
        Overlap: a
        Preutterance: a
        Fixed: a
        Cutoff: a
2. In general this should fit voicebanks recorded with 1 syllable = 1 beat in the specified BPM. If the voicebank is recorded with syllables having different numbers of beats, like a single sample with some 2-beat syllables and some 1-beat syllables, the offsets have to be manually adjusted after generating. But I'm not aware of anyone besides me who does this, and I write my OTOs by hand if the timing is really unusual.
3. The VCV generator was included by accident but I figured I'd describe it anyway instead of removing it completely. I can still remove it from this PR, so that this one is focused on the CVVC generator only.

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/plugin.json:
```diff
@@ -144,75 +144,32 @@
-          "en": "CV Preutterance (ms)",
-          "zh": "CV 的先行发声（毫秒）",
-          "ja": "CV の先行発声（ミリ秒）"
+          "en": "Initial vowel stabilization (ms)"
```
Please add translations for zh and ja using translators. If you find it difficult I can do it later after merging.

adlez27 replied:
I can try adding the Japanese ones.

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
-function pushCV(sample, index, alias, isSpecial, tag) {
-    // check alias count
-    let max = repeat
+function checkAliasCount(alias, isOther, isSingleC) {
```
because it returns the alias, maybe `getAliasWithCount`?

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
     let notes = new Notes()
     if (appendTags) {
-        notes.tag = tag
+        notes.tag = "Head cv"
```
`Head CV`?

adlez27 replied:
This is lowercase so that you can search for "tag:CV" in the entry list without including head cv.

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
+    
+    let end = start - cutoff
     let points = [fixed, preu, ovl]
-    // for oto labeler plus, adding start again in the points
```
This is quite outdated and thank you for removing it 👍 
So I think we can just do `let points = [fixed, preu, ovl, start]`?

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
-    let preu = start + preuVC
+    let start = offset + index * beatLength - consLength - 2*ovlVC
     let ovl = start + ovlVC
+    let preu = start + 2*ovlVC
```
[Very nit] Do you have a JS formatter?

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
+    let extras = [cutoff.toString()]
+    let notes = new Notes()
+    if (appendTags) {
+        notes.tag = "vcv"
```
`VCV`?

adlez27 replied:
This is lowercase so that you can search for "tag:CV" in the entry list without including vcv entries.

sdercolin replied:
I see, this is reasonable.
Seems we need more development on the filter feature 😢

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
     }
 
-    let rest = (sample + appendSuffix).slice(prefix.length)
+    // Remove ".wav" extension before appending suffix
+    let rest = (sample.slice(0,-4) + appendSuffix).slice(prefix.length)
```
We now support not only wav files, so please use [this API](https://github.com/sdercolin/vlabeler/blob/main/docs/file-api.md#related-global-functions).

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
+        }
+        let nextMatched = texts.find(text => next.startsWith(text))
+        if (nextMatched !== undefined) {
+            let nextCons = map.get(nextMatched)[0]
```
What's this `map` (I couldn't find where it's passed)? Do we need to check if `map.get(nextMatched)` has a value before getting its first element?

adlez27 replied:
`map` is created at Ln 78, and since `texts` is created from `map`, `map.get(nextMatched)` will definitely have a value when `nextMatched` is defined.

sdercolin replied:
I see. Could you give it a meaningful name?

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/cvvc-oto-gen.js:
```diff
@@ -117,136 +112,322 @@ let aliasCountMap = new Map()
+        let nextMatched = texts.find(text => next.startsWith(text))
+        if (nextMatched !== undefined) {
+            let nextCons = map.get(nextMatched)[0]
+            if (nextCons !== "") {
```
We can use just `if (nextCons)` to make sure it's not `undefined`, `null`, or an empty string.

sdercolin commented on resources/common/plugins/template/cvvc-oto-gen/plugin.json:
```diff
@@ -1,13 +1,13 @@
 {
   "name": "cvvc-oto-gen",
-  "version": 8,
+  "version": 9,
```
Could you check that the parameters saved in the previous version can be properly loaded in the new version?
- Removed parameters should be ignored
- New parameters should be using default values

adlez27 replied:
Looks like everything happens as expected when using the new version. Unchanged parameters still keep their saved values and new parameters initalize with the default values.

sdercolin reviewed (commented):
The approach LGTM with some minor comments.

I actually think we can just replace the VCV generator as well (directly edit it), if you are confident users probably do not need the separated inputs for each parameter.

sdercolin reviewed (approved):
LGTM! I will add some texts later
